src/Compilers/Core/Portable/DiagnosticAnalyzer/ShadowCopyAnalyzerAssemblyLoader.cs

```diff
@@ -165,37 +166,71 @@ string copyAnalyzerContents()
                 var analyzerFileName = Path.GetFileName(originalAnalyzerPath);
                 var shadowDirectory = Path.Combine(_shadowCopyDirectoryAndMutex.Value.directory, mvid.ToString());
                 var shadowAnalyzerPath = Path.Combine(shadowDirectory, analyzerFileName);
-                copyFile(originalAnalyzerPath, shadowAnalyzerPath);
+                CopyFile(originalAnalyzerPath, shadowAnalyzerPath);
 
-                if (cultureNames.IsEmpty)
+                return shadowAnalyzerPath;
+            }
+        }
+
+        protected override void PrepareSatelliteAssemblyToLoad(string originalAnalyzerPath, string cultureName)
+        {
+            var mvid = AssemblyUtilities.ReadMvid(originalAnalyzerPath);
+            if (_mvidSatelliteAssemblyPathMap.TryGetValue((mvid, cultureName), out Task? copyTask))
+            {
+                copyTask.Wait();
+                return;
+            }
+
+            var tcs = new TaskCompletionSource<bool>();
+            var task = _mvidSatelliteAssemblyPathMap.GetOrAdd((mvid, cultureName), tcs.Task);
+            if (object.ReferenceEquals(task, tcs.Task))
```

Comment:
_Non-blocking_: The management of the `Task` here is virtually the same between the two methods. Feels like a case where we should be able to write it once and use it in both places. Don't have a lot of great recommendations though on best way to do this.

Reply:
I had doubts, but the merging ended up being pretty clean after I changed the Task types to match